lib/omnibus/packager.rb

```diff
@@ -30,24 +28,131 @@ module Packager
     autoload :IPS,      'omnibus/packagers/ips'
     autoload :RPM,      'omnibus/packagers/rpm'
 
-    #
-    # The list of Ohai platform families mapped to the respective packager
-    # class.
-    #
-    # @return [Hash<String, Class>]
-    #
-    PLATFORM_PACKAGER_MAP = {
-      'debian'   => DEB,
-      'fedora'   => RPM,
-      'suse'     => RPM,
-      'rhel'     => RPM,
-      'wrlinux'  => RPM,
-      'aix'      => BFF,
-      'solaris'  => Solaris,
-      'ips'      => IPS,
-      'windows'  => [MSI, APPX],
-      'mac_os_x' => PKG,
-    }.freeze
+    class Platform
+      require 'omnibus/packagers/base'
+
+      def initialize(platform_information)
+        @platform_info = platform_information
+      end
+
+      def self.create(platform_information)
+        platform_name = platform_information['platform_family']
+
+        begin
+          name = Packager.const_get(class_name(platform_name))
+        rescue NameError
+          name = DefaultPlatform
+        end
+        name.new(platform_information)
+      end
+
+      # This function should be overritten by its child class
+      def supported_packagers
+        raise Exception.new("Not implimented")
+      end
+
+      protected
+      def satisfies_version_constraint?(version_constraint)
+        version = @platform_info['platform_version']
+        Chef::Sugar::Constraints::Version.new(version).satisfies?(version_constraint)
+      end
+
+      private
+      def self.class_name(platform_name)
+        "#{platform_name.capitalize}Platform"
+      end
+    end
+
+    class DefaultPlatform < Platform
+      include Logging
+      require 'omnibus/packagers/makeself'
+
+      def supported_packagers
+        platform = @platform_info['platform_family']
+        log.warn(log_key) do
+          "Could not determine packager for `#{platform}', defaulting to `makeself'!"
+        end
+
+        [Makeself]
+      end
+    end
+
+    class Solaris2Platform < Platform
+      require 'omnibus/packagers/solaris'
+      require 'omnibus/packagers/ips'
+
+      def supported_packagers
```

Comment:
What happens if we encounter an unsupported version less than 5.10? I'm assuming we end up failing to build because we don't have a packager and so it would be better to have that failure happen here.

Style-wise, I'd prefer an if-as-expression style over the multiple `return`s.